probe/proc/cache.go

```diff
@@ -0,0 +1,132 @@
+package proc
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+const (
+	filesCacheLen        = 1024
+	filesCacheExpiration = 60 * time.Second
+)
+
+var (
+	errInvalidCacheEntry = errors.New("invalid cache entry")
+)
+
+type filesCacheEntry struct {
+	sync.RWMutex
+	file     File
+	contents *bytes.Buffer
+	creation time.Time
+}
+
+func (fce *filesCacheEntry) Pinned() bool { return fce.contents != nil }
+func (fce *filesCacheEntry) ReadInto(buf *bytes.Buffer) error {
+	fce.Lock()
+	defer fce.Unlock()
+
+	if fce.Pinned() { // use the saved contents
+		buf.Write(fce.contents.Bytes())
+		return nil
+	}
+	return fce.file.ReadInto(buf)
+}
+
+// A cache for files handles
+type filesCache struct {
+	*lru.Cache
+	proc Dir
+	quit chan struct{}
+}
+
+func newFilesCache(proc Dir) *filesCache {
+	evictionFunc := func(key, value interface{}) {
+		entry, ok := value.(filesCacheEntry)
+		if ok && entry.file != nil {
+			entry.file.Close()
+		}
+	}
+
+	cache, err := lru.NewWithEvict(filesCacheLen, evictionFunc)
+	if err != nil {
+		panic("could not create cache")
+	}
+	fc := filesCache{
+		Cache: cache,
+		proc:  proc,
+		quit:  make(chan struct{}),
+	}
+	go fc.loop()
+
+	return &fc
+}
+
+// Read a "/proc" file, identified as a sub-path (eg "1134/comm"), into a buffer
+// When the "pin" argument is true, we will cache the file _content_.
+func (fc *filesCache) ReadInto(filename string, buf *bytes.Buffer, pin bool) error {
+	e, ok := fc.Cache.Get(filename)
+	if !ok {
+		f, err := fc.proc.Open(filename)
+		if err != nil {
+			return err
+		}
+		entry := filesCacheEntry{
+			creation: time.Now(),
+		}
+		b := bytes.Buffer{}
+		if err := f.ReadInto(&b); err != nil {
+			return err
+		}
+		if pin {
+			entry.contents = &b // save the file contents
+			f.Close()           // and close the file
+		} else {
+			entry.file = f
+		}
+		buf.Write(b.Bytes())
+		fc.Cache.Add(filename, entry)
```

Comment:
What happens to the other filesCacheEntry if this function is called concurrently?  You'll end up calling add twice, which will probably loose all the references to the fce, leaking open files.